File: android/src/main/java/com/rnmapbox/rnmbx/modules/CustomHttpHeaders.kt
package com.rnmapbox.rnmbx.modules

import com.mapbox.common.*
import com.rnmapbox.rnmbx.v11compat.httpinterceptor.*

data class CustomHttpHeadersOptions(val urlRegexp: Regex?)

data class CustomHttpHeadersMapValue(
    val headerValue: String,
    val options: CustomHttpHeadersOptions?
)

object CustomHttpHeaders : HttpServiceBase() {
    const val LOG_TAG = "CustomHttpHeaders"

    init {}

    val map = mutableMapOf<String, CustomHttpHeadersMapValue>()

    fun addCustomHeader(headerName: String, headerValue: String, options: CustomHttpHeadersOptions? = null) {
        HttpServiceFactory.getInstance().setInterceptor(
            this
        )
        map.put(headerName, CustomHttpHeadersMapValue(headerValue = headerValue, options = options))
    }

    fun removeCustomHeader(headerName: String) {
        map.remove(headerName)
    }

    fun getCustomRequestHeaders(customRequestHeaders: MutableMap<String, CustomHttpHeadersMapValue>, httpRequest: HttpRequest): HashMap<String, String> {
        val headers = hashMapOf<String, String>()
        for (entry in map.entries.iterator()) {
            val urlRegexp = entry.value.options?.urlRegexp
            if (urlRegexp != null) {
                val destination = httpRequest.headers.getOrDefault("location", httpRequest.url)
                if (urlRegexp.matches(destination)) {
                    headers[entry.key] = entry.value.headerValue
                }
            }
            else {
                headers[entry.key] = entry.value.headerValue
            }
        }
        return headers
    }

    override fun onRequest(request: HttpRequest): HttpRequest {
        request.headers.remove("Authorization")
        request.headers.putAll(getCustomRequestHeaders(map, request))
        return request
    }

    override fun onDownload(download: DownloadOptions): DownloadOptions {
        download.request.headers.putAll(getCustomRequestHeaders(map, download.request))
        return download
    }

    override fun onResponse(response: HttpResponse): HttpResponse {
        return response
    }
}

File: ios/RNMBX/CustomHttpHeaders.swift
import MapboxMaps

struct CustomHttpHeadersOptions {
  var urlRegexp: NSRegularExpression?
}

struct CustomHttpHeadersMapValue {
  var headerValue: String
  var options: CustomHttpHeadersOptions
}

class CustomHttpHeaders : HttpServiceInterceptorInterface {
  func onRequest(for request: HttpRequest, continuation: @escaping HttpServiceInterceptorRequestContinuation) {
    let request = onRequest(for: request)
    continuation(HttpRequestOrResponse.fromHttpRequest(request))
  }

  func onResponse(for response: HttpResponse, continuation: @escaping HttpServiceInterceptorResponseContinuation) {
    continuation(response)
  }

  static var shared : CustomHttpHeaders = {
    let headers = CustomHttpHeaders()
    headers.install()
    return headers
  }()

  var customHeaders : [String:CustomHttpHeadersMapValue] = [:]

  func install() {
    HttpServiceFactory.setHttpServiceInterceptorForInterceptor(self)
  }

  func reset() {
    HttpServiceFactory.setHttpServiceInterceptorForInterceptor(nil)
  }

  // MARK: - HttpServiceInterceptorInterface
  
  func getCustomRequestHeaders(for request: HttpRequest, with customHeaders: [String: CustomHttpHeadersMapValue]) -> [String: String] {
    var headers: [String: String] = [:]
    let urlString = request.url

    for (key, entry) in customHeaders {
      let options = entry.options

      if let pattern = options.urlRegexp {
        do {
          let range = NSRange(location: 0, length: urlString.utf16.count)
          if pattern.firstMatch(in: urlString, options: [], range: range) != nil {
            headers[key] = entry.headerValue
          }
        }
      } else {
        headers[key] = entry.headerValue
      }
    }
    return headers
  }


  func onRequest(for request: HttpRequest) -> HttpRequest {
    let customHeaders = getCustomRequestHeaders(for: request, with: customHeaders)
    request.headers.merge(customHeaders) { (_, new) in new }
    return request
  }

  func onResponse(for response: HttpResponse) -> HttpResponse {
    return response
  }
}
